lara-typescript/src/page-item-authoring/common/hooks/use-authoring-preview.ts

```diff
@@ -0,0 +1,72 @@
+import { ILinkedInteractive, ISetLinkedInteractives } from "@concord-consortium/interactive-api-host";
+import { RefObject, useCallback, useEffect, useState } from "react";
+import { debounce } from "ts-debounce";
+
+interface IInteractiveBase {
+  url?: string | null;
+  authored_state: string;
+  linked_interactives: ILinkedInteractive[];
+}
+
+interface IPreviewUpdate {
+  newAuthoredState: string | object;
+  newUrl?: string | null;
+  newLinkedInteractives?: ILinkedInteractive[];
+}
+
+interface IOptions {
+  interactive: IInteractiveBase;
+  interactiveAuthoredStateRef: RefObject<HTMLTextAreaElement>;
+  linkedInteractivesRef: RefObject<HTMLTextAreaElement>;
+  handleUpdateItemPreview?: (updates: Record<string, any>) => void;
+}
+
+export const useAuthoringPreview = (options: IOptions) => {
+  const { interactive, handleUpdateItemPreview, interactiveAuthoredStateRef, linkedInteractivesRef } = options;
+
+  const [url, setUrl] = useState(interactive.url);
+  const [authoredState, setAuthoredState] = useState(interactive.authored_state);
+  const [linkedInteractives, setLinkedInteractives] =
+    useState<ILinkedInteractive[] | undefined>(interactive.linked_interactives);
+
+  const _updatePreview = ({ newAuthoredState, newLinkedInteractives, newUrl }: IPreviewUpdate) => {
+    handleUpdateItemPreview?.({
+      authoredState: typeof newAuthoredState === "string" ? newAuthoredState : JSON.stringify(newAuthoredState),
+      linkedInteractives: newLinkedInteractives,
+      url: newUrl
+    });
+  };
+
+  const updatePreview = useCallback(debounce(_updatePreview, 500), []);
```

Comment:
This approach of `useCallback(debounce(...), [])` is confusing to me. If it makes sense to you, can you add a comment about how it works?

As far as I understand `debounce` will be called each time this hook is called. However, only the function returned by debounce the first time will be used used due to the dependencies of `[]`.  So updatePreview will always be a debounced version of the first `_updatePreview` function.  If the caller of `userAuthoringPreview` passes a different `handleUpdateItemPreview` function this will never be called.  I might have that all wrong though.

Reply:
Some suggest using `useMemo` for this purpose, e.g. https://dmitripavlutin.com/react-throttle-debounce/.

Reply:
> As far as I understand debounce will be called each time this hook is called. However, only the function returned by debounce the first time will be used used due to the dependencies of []. So updatePreview will always be a debounced version of the first _updatePreview function. If the caller of userAuthoringPreview passes a different handleUpdateItemPreview function this will never be called. 

@scytacki, I think you're right. This line was copied from the old code found in MwInteractive and ManagedInetractive React components. I saw it was working fine (the delay was noticeable and updates were merged), so frankly speaking, I didn't think about it too much while moving it to this hook.

I've just pushed one more commit that cleans it up following some advice from @kswenson's article.

On a related note - it seems that our linter + rules of hooks don't work too well. I believe it should have complained about the original code (empty dependency array) and it didn't. I've noticed that in a few other places too. I bet it's caused by the fact that we use deprecated TSLint + its rules of hooks plugin. I guess it'd be great to migrate to ESLint when we have another time for maintenance.